src/lib.rs

```diff
@@ -0,0 +1,310 @@
+mod baseline;
+mod check_release;
+mod config;
+mod dump;
+mod manifest;
+mod query;
+mod templating;
+mod util;
+
+pub use config::*;
+pub use query::*;
+
+use check_release::run_check_release;
+use trustfall_rustdoc::load_rustdoc;
+
+use std::{collections::HashSet, path::PathBuf};
+
+/// Test a release for semver violations.
+#[derive(Default)]
+pub struct Check {
+    /// Which packages to analyze.
+    scope: Scope,
+    current: Current,
+    baseline: Baseline,
+    log_level: Option<log::Level>,
+}
+
+#[derive(Default)]
+enum Baseline {
+    /// Version from registry to lookup for a baseline. E.g. "1.0.0".
+    Version(String),
+    /// Git revision to lookup for a baseline.
+    Revision(String),
+    /// Directory containing baseline crate source.
+    Root(PathBuf),
+    /// The rustdoc json file to use as a semver baseline.
+    RustDoc(PathBuf),
+    /// Latest version published to the cargo registry.
+    #[default]
+    LatestVersion,
+}
+
+/// Current version of the project to analyze.
+#[derive(Default)]
+enum Current {
+    /// Path to the manifest of the current version of the project.
+    /// It can be a workspace or a single package.
+    Manifest(PathBuf),
+    /// The rustdoc json of the current version of the project.
+    RustDoc(PathBuf),
+    /// Use the manifest in the current directory.
+    #[default]
+    CurrentDir,
+}
+
+#[derive(Default)]
+struct Scope {
+    selection: ScopeSelection,
+    excluded_packages: Vec<String>,
+}
+
+/// Which packages to analyze.
+#[derive(Default, PartialEq, Eq)]
+enum ScopeSelection {
+    /// Package to process (see `cargo help pkgid`)
+    Packages(Vec<String>),
+    /// All packages in the workspace. Equivalent to `--workspace`.
+    Workspace,
+    /// Default members of the workspace.
+    #[default]
+    DefaultMembers,
+}
+
+impl Scope {
+    fn selected_packages<'m>(
```

Comment:
this function is very similar to [partition_packages](https://docs.rs/clap-cargo/latest/clap_cargo/struct.Workspace.html#method.partition_packages).
Initially, I wanted to use it, but since the struct `clap_cargo::Workspace` is non-exhaustive, it cannot be constructed.